rust/hyperlane-base/src/contract_sync/cursor.rs

```diff
@@ -1,90 +1,348 @@
+    /// at which it was dispatched.
+    /// Otherwise, rewind all the way back to the start block.
+    async fn update(&mut self, logs: Vec<(HyperlaneMessage, LogMeta)>) -> eyre::Result<()> {
+        let prev_nonce = self.0.next_nonce.saturating_sub(1);
+        self.0.update(logs, prev_nonce).await
+    }
+}
+
+/// A MessageSyncCursor that syncs backwards to nonce zero.
+#[derive(new)]
+pub(crate) struct BackwardMessageSyncCursor {
+    cursor: MessageSyncCursor,
+    synced: bool,
+}
+
+impl BackwardMessageSyncCursor {
+    async fn get_next_range(&mut self) -> Option<(u32, u32, Duration)> {
+        // Check if any new messages have been inserted into the DB,
+        // and update the cursor accordingly.
+        while !self.synced {
+            if let Some(block_number) = self
+                .cursor
+                .retrieve_dispatched_block_number(self.cursor.next_nonce)
+                .await
+            {
+                self.cursor.next_block = block_number;
+                // If we found nonce zero, we are done rewinding.
+                if self.cursor.next_nonce == 0 {
+                    self.synced = true;
+                    break;
+                }
+                self.cursor.next_nonce = self.cursor.next_nonce.saturating_sub(1);
+            } else {
+                break;
+            }
+        }
+        if self.synced {
+            None
+        } else {
+            // Just keep going backwards.
+            let to = self.cursor.next_block;
+            let from = to.saturating_sub(self.cursor.chunk_size);
+            self.cursor.next_block = from.saturating_sub(1);
+            // TODO: Consider returning a proper ETA for the backwards pass
+            Some((from, to, Duration::from_secs(0)))
+        }
+    }
+
+    /// If the previous block has been synced, rewind to the block number
+    /// at which it was dispatched.
+    /// Otherwise, rewind all the way back to the start block.
+    async fn update(&mut self, logs: Vec<(HyperlaneMessage, LogMeta)>) -> eyre::Result<()> {
+        let prev_nonce = self.cursor.next_nonce.saturating_add(1);
+        self.cursor.update(logs, prev_nonce).await
+    }
+}
+
+enum SyncDirection {
+    Forward,
+    Backward,
+}
+
+/// A MessageSyncCursor that syncs forwards in perpetuity.
+pub(crate) struct ForwardBackwardMessageSyncCursor {
+    forward: ForwardMessageSyncCursor,
+    backward: BackwardMessageSyncCursor,
+    direction: SyncDirection,
+}
+
+impl ForwardBackwardMessageSyncCursor {
+    /// Construct a new contract sync helper.
+    pub async fn new(
+        indexer: Arc<dyn MessageIndexer>,
+        db: Arc<dyn HyperlaneMessageDB>,
+        chunk_size: u32,
+    ) -> Result<Self> {
+        let (count, tip) = indexer.fetch_count_at_tip().await?;
+        let forward_cursor = ForwardMessageSyncCursor::new(MessageSyncCursor::new(
+            indexer.clone(),
+            db.clone(),
+            chunk_size,
+            tip,
```

Comment:
Tip twice feels odd to me

Reply:
I would rather we define `new` by hand because then the knowledge it needs to be duplicated is handled next to the thing that requires knowing it. `new` is just a convenience for certain cases.

Reply:
The amount of code that we'd write vs the amount of knowledge that we'd prevent callers from needing to know doesn't feel like a winning tradeoff to me?